Block API: Add `supports` object to block attributes to configure copy support

stacimc opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->
Closes #29693
Alternative to #32604, #34750

## Description
<!-- Please describe what you have changed or added -->

This PR:
* Adds handling for a new `supports` object property to block attributes. It includes the `copy` support, used to indicate whether the attribute should be copied on block duplication. It is true by default.

Example fragment of block.json:

```
            "attributes": {
		"myAttributeThatShouldNotBeCopied": {
			"type": "string",
			"default": "my-default-value",
			"__experimentalSupports": {
				"copy": false,
			}
		}
	},
```
Copy support is true by default, meaning it is true if the `supports.copy` is not configured, and if the `supports` property is missing altogether.
* Refactors `__internalWidgetID` to use this new property/
 

For just one example of how this might be used: the Jetpack `Pay with Paypal` block uses a `productId` attribute to identify the actual product record referenced by the block. This data is meant to be internal to the block and should not be exposed to the user. When the user duplicates the block, we want to copy the other attributes used to store price info and other details, but create a brand new product record/productId.


## Notes

* In other regards, an attribute configured this way acts like a normal attribute
    * ~~It respects any configured default values~~ Default values are not filled in on block duplication either; support for this could be easily added if a viable use case becomes apparent, though
    * It can be managed as normal via `setAttributes`
    * It can apply to any type of attribute (ie it is not necessarily an id field)
* While you can give this role to an attribute with a "source"/"selector" configuration, it is not meaningful to do so. For example, adding it to the `content` attribute of the [Code block](https://github.com/WordPress/gutenberg/blob/a84c037a0e62a344005054102544c34d7b970a6b/packages/block-library/src/code/block.json#L9) would not make sense, because although the attribute would not be copied on duplication, it will populate with the same value from the html.
* Because this is experimental and also relies on experimental features, I have not added documentation; happy to do so if that is appropriate!
* A note about Reusable blocks:
Attributes without copy support are not removed when you convert a block into a Reusable block. I don’t know a ton about Reusable blocks, so calling this out so that someone more familiar can correct me; my understanding is that multiple instances of a Reusable block are not independent of each other (ie editing one copy affects them all), so even 'internal' attributes should also be shared.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
* Play around with adding `supports: { copy: false }` to any block attribute. 
* Insert the block into a new post, change its attributes, and duplicate it
* Verify that the un-copyable attributes were not copied, but all other attributes were
* ~~Test with attributes that have a `default` value and ensure that the default is respected in the duplicated block~~
* Test Copy+Paste using both the `Copy` menu item, and by keyboard commands
* Test duplicating a block in the widget editor
* Test adding and editing widgets in both the Widget editor and the Customizer
  * Test moving widgets between widget areas/side panels
  * Test a server-side rendered block in the widget editor and make sure the preview is displayed correctly (Basically verify that #29197 is still fixed with the new implementation)

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

stacimc committed:
Add internal experimentalRole for attributes

stacimc committed:
Add tests

stacimc committed:
Refactor removeAttributesByRole to take role param and return block

* For extensibility, it can now be used to remove attributes with any
given role
* It takes and returns a block instance; this is so that it will be
easier to use in copy handlers/etc

stacimc committed:
Remove internal attrs when copying via keyboard command

stacimc committed:
Remove internal attrs when copying via CopyMenuItem

stacimc committed:
Do not alter attributes of block being duplicated

`__experimentalRemoveAttributesByRole` takes a block instance and removes
any attributes with the given `role`. When cloning a block, we want to
remove internal attributes on the clone, but not the original copy. This
changes the order of operations in the cloning function to make sure that
those attributes only get removed on the latter.

Without this change, when a block is duplicated in the widget editor all
internal attributes will be stripped from BOTH the copy and the duplicate.

stacimc committed:
Use `__experimentalCloneSanitizedBlock` in copy handlers

When a block is duplicated, we remove internal attributes and then fill in
any default values (when we call `__experimentalSanitizeBlockAttributes`).
When we copy a block, we were only doing the first step of removing the
internal attrs, but not filling the defaults.

This causes a block validation issue, because the serialized block in the
clipboard is missing internal attrs which will get filled in by `createBlock`
when we actually paste.

This commit updates the copy handlers to use the cloneBlock code so that the
behavior matches block duplication exactly. It also reverts the
`__experimentalRemoveAttributesByRole` to the original API that operates on
an attributes object rather than a block instance, to match the signature of
`__experimentalSanitizeBlockAttributes` and clean up the logic in the
cloning function.

stacimc committed:
Use internal attribute for widgetId

stacimc committed:
Update tests for __experimentalRemoveAttributesByRole

ntsekouras committed:
Update packages/widgets/package.json

stacimc committed:
Move attribute registration to function called in widget initialization

stacimc committed:
Corrected docstring

stacimc committed:
Check for existence of widget before moving sidebars

When you move a block in the customizer to a different widget sidebar,
we check to see if the block has a widgetId and if so, assume it's a
valid widget that we can just remove from the original sidebar and add
to the destination sidebar. If it's an inner block of a widget, we
instead remove the block from the parent, create a widget from the block,
and add this to the new sidebar. To detect this we expect it to not have a
wwidgetId.

This can break when you take an existing widget and turn it into inner
blocks by transforming it into a Group. The inner block retains its
widgetId through this transformation, but the corresponding widget
instance is removed.

To cover this case, we now check both that the widgetId exists AND that
there is an existing widget with that id.

stacimc committed:
Do not separately sanitize attributes in cloneSanitizedBlock

The only reason we still had this was to fill in default values for
any internal attrs that got stripped on duplication; since there is
no obvious use case for an internal attribute with defaults I'm
removing this for now.

I leave __experimentalSanitizeBlockAttributes as it is still used in
createBlock, and keeping it pulled out like this allows for ease of
unit testing.

stacimc committed:
Remove __experimentalSanitizeBlockAttributes

stacimc committed:
Remove unnecessary attribute sanitization in ServerSideRender component

This was previously added because the endpoint called by this component
validates attributes, which would break with the previous implementation of
`__internalWidgetId` (since it wasn't a registered attribute).

With the new implementation, `__internalWidgetId` is a valid attribute and
it's no longer necessary to sanitize here.

stacimc committed:
Remove removeAttributesByRole

stacimc committed:
Remove experimentalCloneSanitizeBlock and make retaining internal attrs an option for cloneBlock

stacimc committed:
Add option to serialize to retain internal attrs

stacimc committed:
Refactor to use a 'supports' property instead of '__experimentalRole'

stacimc committed:
Do not pass widgetId to /block-renderer

stacimc committed:
Update test

stacimc commented:
Relative to the implementation proposed in #34750, these are the changes, (thanks @noisysocks 😄):

* The API used to make use of the `__experimentalRole` to create 'internal' attributes. This PR instead adds a `supports` object:
```
	"attributes": {
		"myAttributeThatShouldNotBeCopied": {
			"type": "string",
			"default": "my-default-value",
			"supports": {
				"copy": false,
			}
		}
	},
```
* Removes `__experimentalCloneBlock` and instead provides an option for both `cloneBlock` and `serialize` to toggle retention of non-copyable attributes (by default they _will_ be copied):
```
const serialized = serialize( blocks, { retainCopyAttributes: false  } );
const cloned = cloneBlock( block, {}, null, { retainCopyAttributes: false } )
```
* The pre-existing `__experimentalGetBlockAttributesNamesByRole` utility has been modified to `__experimentalFilterBlockAttributes` and now filters attributes by arbitrary properties, so:
```
__experimentalFilterBlockAttributes( blockName, { __experimentalRole: 'content' } );
__experimentalFilterBlockAttributes( blockName, { supports: { copy: false } } );
```

stacimc commented on packages/block-editor/src/components/block-settings-menu/block-settings-dropdown.js:
```diff
@@ -33,7 +33,10 @@ const POPOVER_PROPS = {
 function CopyMenuItem( { blocks, onCopy } ) {
-	const ref = useCopyToClipboard( () => serialize( blocks ), onCopy );
+	const ref = useCopyToClipboard(
+		() => serialize( blocks, { retainCopyAttributes: false } ),
```
`retainCopyAttributes` isn't a spectacular name, but I had difficulty coming up with a better one. `excludeNonCopyableAttributes`? `retainOnlyCopyableAttributes`?

noisysocks replied:
Maybe `includeCopyableAttributes`? Also let's add `__experimental` for the moment 😀

stacimc replied:
Agreed about `__experimental` :)

Updated to `__experimentalExcludeNonCopyableAttributes` (and inverted the logic). It's an... intense name 😅 but I'm having trouble being more concise. I'm worried that something like `retainCopyAttributes` or `includeCopyableAttributes` doesn't really get across that it means retain _only_ copyable attributes.

stacimc committed:
Update use of 'internal' terminology in docs

stacimc commented on packages/server-side-render/src/server-side-render.js:
```diff
@@ -88,19 +87,17 @@ export default function ServerSideRender( props ) {
+		// The __internalWidgetId is only defined on the client-side and must be removed before making the
+		// request.
+		const retainedAttributes = omit( attributes, [ '__internalWidgetId' ] );
 
```
We might be able to do this in a more generic way, but the `__internalWidgetId` is actually a special case and is the only attribute that needs to be stripped here 🤔 

My plan was to just add back in `__experimentalSanitizeBlockAttributes`, which on trunk handles the removal of the widgetId -- but that will no longer work! That function removes _un-registered_ attributes, but the widgetId _is_ registered now (just only on the client side).

We _don't_ want to omit all attributes with `copy: false` here, because it's entirely possible to add 'normal' server-side attributes that don't support copy, and those shouldn't be stripped. Only the widgetId is a special case.

Alternative suggestions from @noisysocks on #34750:
> We could just set 'additionalProperties' to true on the server. I don't really see the need to be so strict there. This might require a Core patch, though, as I'm not sure that it's possible for a plugin to re-register a route.
>
> Or, we could have ServerSideRender perform a GET on the /block-types endpoint before making its request to /block-renderer. If an attribute isn't in the attributes object that comes back from /block-types, then it hasn't been defined on the server, and so we must not send it to /block-renderer.

noisysocks replied:
> We could just set 'additionalProperties' to true on the server. I don't really see the need to be so strict there. This might require a Core patch, though, as I'm not sure that it's possible for a plugin to re-register a route.

I'm leaning towards this since it matches how the frontend works, i.e. you can put whatever you like in the block's markup and the block editor will silently ignore it.

But yeah this is fine as a temporary solution. Before merging, could you please create an issue to track this and link to it in a `// TODO`?

noisysocks commented on packages/blocks/src/api/factory.js:
```diff
@@ -94,62 +124,45 @@ export function createBlocksFromInnerBlocksTemplate(
+ * @param {Object}   block                        Block instance.
+ * @param {Object}   mergeAttributes              Block attributes.
+ * @param {?Array}   newInnerBlocks               Nested blocks.
+ * @param {?Object}  options                      Cloning options.
```
I wonder if it makes sense to mark this as `__experimental`. It would be easy to miss unless the developer calling the function was looking closely at the docs. At any rate, we have until WP 6.0 to decide to remove or change it.

noisysocks commented:
I much prefer this 😀 it's "KISS" and we're not overloading the meaning of `role` which is for giving an attribute _semantic_ meaning. I'll look more closely at and test the PR on Monday.

draganescu commented:
I think this is better than `internal` which does not hold any meaning and could be interpreted in any way.

stacimc committed:
Mark supports object as experimental

stacimc committed:
Update option name, mark as experimental

stacimc committed:
Add TODO to remove special handling for internalWidgetId when possible

stacimc commented:
Updated to change `supports` --> `__experimentalSupports`, as well as the new option to `cloneBlock`.

dmsnell commented on packages/block-editor/src/components/copy-handler/index.js:
```diff
@@ -122,7 +122,10 @@ export function useClipboardHandler() {
-				const serialized = serialize( blocks );
+				const serialized = serialize( blocks, {
+					__experimentalExcludeNonCopyableAttributes:
+						event.type === 'copy',
```
this one feels overly specific when we're introducing a general way we could accomplish this too.

```ts
const serializeOptiosn = event.type === 'copy'
	? { __experimentalIncludeAttributes: [ 'copy' ]
	: undefined;

const serialized = serialize( blocks, serializeOptions );
```

---

Also I'm really lost on the double negative. A positive phrasing could make that easier to understand.

dmsnell replied:
(this would give us the ability to denote multiple sets of attributes for inclusion _or_ sets for explicit exclusion)

```ts
{
	__experimentalExcludeAttributes: [ 'style' ]
}
```

stacimc replied:
The variable name definitely leaves a _lot_ to be desired, suggestions are very much appreciated 😄

Making it extensible to other sets of attributes is great! It gets a little busy if we want to include the ability to filter out by _any_ attribute property (rather than limiting it to a list of `supports` keys for example):

```
const serializationOptions =  {
    __experimentalIncludeAttributes: {
            __experimentalSupports: { copy: false },
    }
}
const serialized = serialize(
    blocks,
    event.type === 'copy' ? serializationOptions : undefined
);
```

Not sure much can be done about that 🤔 I think the benefit to extensibility and the clearer naming is a big win.

dmsnell replied:
just a note about my example: I wasn't proposing nesting `supports` inside of `includeAttributes`. using your object notation instead of an array it could be this too, which is closer to what I was trying to imply. the include/exclude arrays have some advantages, but they aren't directly necessary here either

```ts
const options = {
	__experimentalAttributeSupports: {
		copy: false,
	}
}
```

stacimc replied:
This is tricky. Nesting `supports` inside of `includeAttributes` is very verbose but reads clearly to me. I'm a little concerned with this:

```
// Is it clear from this that attributes that do not support `copy` will be omitted?
cloneBlock(
    block,
    __experimentalOptions: {
        __experimentalSupports: { copy: false }
    }
}

// Would it be easy to confuse with this -- here attributes that DO support `copy` would be omitted:
cloneBlock(
    block,
    __experimentalOptions: {
        __experimentalSupports: { copy: true }
    }
}
```

Sorry if I'm misunderstanding, the double negatives are confusing. 

```
// This is more 'streamlined' and reads more clearly to me, but only supports excluding attributes by `supports` keys. Eg you can't exclude attributes by their role for example
cloneBlock(
    block,
    __experimentalOptions: {
        __experimentalExcludeAttributes: { copy: false }
    }
}
```

dmsnell commented on packages/block-editor/src/store/actions.js:
```diff
@@ -1172,7 +1171,7 @@ export const duplicateBlocks = ( clientIds, updateSelection = true ) => ( {
 	);
 	const clonedBlocks = blocks.map( ( block ) =>
-		__experimentalCloneSanitizedBlock( block )
+		cloneBlock( block, {}, null, { retainCopyAttributes: false } )
```
wouldn't cloning imply retaining?

same question as above with regards to being overly-specific to `copy`

stacimc replied:
I missed updating `retainCopyAttributes` here, I'll update that when I settle on the API.

> wouldn't cloning imply retaining?

In the driving use cases of a `productId` or `widgetId`, we don't want it retained on `copy` (whether that be from the block menu or keyboard commands) or on duplication. With this API though we could consider those to be separate supports:
```
"__experimentalSupports": {
    "copy": false,
    "duplicate": false
}
```

In either case, I think it's a good point that adding the logic to `cloneBlock` is unnecessarily confusing -- we could clone the block and strip the attributes afterward, here in the duplication logic. In an earlier iteration it made a little more sense to add the logic to `cloneBlock` because we were using it in multiple places.

dmsnell commented on packages/blocks/src/api/serializer.js:
```diff
@@ -354,7 +365,16 @@ export function serializeBlock( block, { isInnerBlocks = false } = {} ) {
+		saveAttributes = omit(
+			saveAttributes,
+			__experimentalFilterBlockAttributes( blockName, {
+				__experimentalSupports: { copy: false },
```
this here feels like it could be the thing exposed through `serializeBlock`'s options. do we need both the inner and outer language of talking about it?

ntsekouras commented:
@stacimc thank you for all your work and the patience with all the back and forth in the approach 😄

Is the PR ready for review? I've started testing(without looking at the code yet) but I'm not observing the expected results.. I've tried adding to `level` and `textAlign` in Heading block.. Also can you please rebase the PR?

stacimc committed:
Correctly exclude noncopyable attrs in duplicate

stacimc commented:
> @stacimc thank you for all your work and the patience with all the back and forth in the approach 😄
> 
> Is the PR ready for review? I've started testing(without looking at the code yet) but I'm not observing the expected results.. I've tried adding to `level` and `textAlign` in Heading block.. Also can you please rebase the PR?

Rebased -- and fixed a bug with `duplicate` that you may have been seeing. I am also planning on addressing @dmsnell's feedback [here](https://github.com/WordPress/gutenberg/pull/38643#discussion_r806336142) now it seems we may be going with this approach.

ntsekouras commented:
>Rebased -- and fixed a bug with duplicate that you may have been seeing.

Thanks so much for doing this! I'll test and review tomorrow!

stacimc committed:
Use __experimentalExcludeAttributes option

stacimc committed:
Use new serializer options

stacimc commented:
Rebased again and an update to the serialization options:

```
serialize( blocks, {
    __experimentalExcludeAttributes: { copy: false },
} )
```

ntsekouras commented on packages/blocks/src/api/serializer.js:
```diff
@@ -354,7 +362,17 @@ export function serializeBlock( block, { isInnerBlocks = false } = {} ) {
+	let saveAttributes = getCommentAttributes( blockType, block.attributes );
+
+	if ( __experimentalExcludeAttributes ) {
+		saveAttributes = omit(
```
Here there is a problem when attributes produce different html like css classes. With my tests in Heading and `textAlign`, a css class is added and produces:
```
<!-- wp:heading {"textAlign":"center"} -->
<h2 class="has-text-align-center">hello</h2>
<!-- /wp:heading -->
```
So we can't use the initial `saveContent` produced by the previous attributes, because it will result in skipping the unwanted attributes but will keep the html. By doing that when we copy we end up with:
```
<!-- wp:heading {"className":"has-text-align-center"} -->
<h2 class="has-text-align-center">hello</h2>
<!-- /wp:heading -->
```

So what we need to do here if we have to omit attributes is:
1. get the remaining attributes by using `block.attributes`
2. get `saveContent` again with the above result
3. recalculate `saveAttributes`

The above flow is for `copy` actions

dmsnell replied:
the serializer is probably not the right place to do this. it's going to cause a lot of unintended consequences elsewhere because of the mashing of concerns. we should be able to exclude the attributes when cloning and then we won't have to touch `serializeBlock` and we won't have any issues with the contradiction between the serialized markup and the block attributes.

dmsnell replied:
do we need this change at all given that it's added already in `cloneBlock()`?

stacimc replied:
Thank you, this is a very good catch 😬 

This was done to support the `copy` actions, so that's:
* using the `Copy` item in the block [toolbar](https://github.com/WordPress/gutenberg/pull/38643/files#diff-9e90ff9d1914908a6c9c3b91a99ad894998ed3b02d4bb466d53d41d8d58f2d99)
* Copy/pasting using the [keyboard](https://github.com/WordPress/gutenberg/pull/38643/files#diff-d6ad6b9ae7f870e6d8d163e142ababc68c51e42faf92405fba08a3d063bb8eeb)

In a previous iteration I was using `cloneBlock` in that logic to remove unsupported attributes (instead of adding the logic to the serializer), but there was [some concern](https://github.com/WordPress/gutenberg/pull/34750#discussion_r760826441) that we'd have to ensure that logic was utilized any time new functionality was added, particularly by plugins. Given that we're relying on passing a prop to the serializer to exclude those attributes though, I don't see that it's very different. I do think it makes sense to set an expectation that you should explicitly `cloneBlock` a block before duplicating it. 

I'll remove the serializer changes.

ntsekouras reviewed (commented):
Thanks again for all your work @stacimc! I started testing the PR with `"__experimentalSupports": { "copy": false }` to `Heading` block in `textAlign` attribute.

The main issue I see is the comment below and I'm not sure if the `duplicate` action can have any implications with `originalContent` property. When we duplicate we insert new blocks and they are rendered properly but they still keep the `stale` originalContent property. @dmsnell  might have some input here..

stacimc committed:
Remove logic from serializer in favor of using cloneBlock in copy handlers

dmsnell reviewed (commented):
I'm a little concerned that our language introduces double negatives: we have `exclude` and `{ copy: false }` which sounds like it's saying, "don't exclude the `copy` attributes" but I think we're actually saying, "go ahead and exclude the `copy` attributes."

Did we consider `__experimentalRetainAttributes: { copy: false }`?
Can we find a good way to communicate this in the affirmative sense, or at least while avoiding the double-negative?

dmsnell commented:
Do we have known use-cases waiting for this on the server-side? Curious what folks plan on doing to clone blocks within a post from the server.

stacimc commented:
Rebased.

> I'm a little concerned that our language introduces double negatives: we have exclude and { copy: false } which sounds like it's saying, "don't exclude the copy attributes" but I think we're actually saying, "go ahead and exclude the copy attributes."
>
> Did we consider __experimentalRetainAttributes: { copy: false }?

@dmsnell The [initial version](https://github.com/WordPress/gutenberg/pull/38643#discussion_r802122415) was something close to that! One concern I have with the word `retain` is that it doesn't really reflect that we are _only_ retaining those attributes, and excluding others. I do agree that the current language could be confusing/misleading, though 😞 

I think part of the problem is that the object passed in is actually a _filter_ which matches the structure of the attribute's configuration in `block.json`. So `__experimentalExcludeAttributes: { copy: false }` should be read not as "don't exclude copy attributes" but as "exclude attributes that match the filter: {copy: false}". That's not obvious.

What do you think about `__experimentalFilterAttributes`? `__experimentalExcludeAttributesByFilter`?

dmsnell commented:
There are a few ways we can address the language, some of which none of us have considered yet.

For one, we don't have to pass an object. Do we ever have a valid case to call `excludeAttributes: { copy: true }` if indeed it implies the double-negative? We could pass an array of attribute sets to exclude and that would remove the double-boolean.

```ts
excludeAttributes: [ 'copy' ]
```

For two I'm not sure that `retain` carries the same heavy implication you mention given that we're in `cloneBlock`, but I see your point.

`filter` seems even more confusing to me.

`__experimentalOmitting: { copy: true }`
`__experimentalExcludeAttributes: { copy: true }`
`__experimentalOmittingAttributes: [ 'copy' ]`

the word `exclude` isn't as troublesome as saying that `exclude: false` implies that we do exclude.

stacimc commented:
`excludeAttributes: [ 'copy' ]` does read "exclude copy attributes" to me, but I'm less sure that "copy attributes" is clearly understood to be attributes that _do not support_ the copy operation. I would likely read that as "exclude copyable attributes". Maybe that's just me, or maybe given that this is only being used in `cloneBlock` (since I removed it from `serialize`), it's no longer likely to be misinterpreted.

The reason I said `filter` is because the object passed is literally a filter which `cloneBlock` uses to[ filter through the block's attributes](https://github.com/WordPress/gutenberg/pull/38643/files#diff-3fbd4c0db6cff0321f313fd5d26d509a897408e2af0d95b488e3fa8c2fb66e8bR157), and omits those that match. This would be clearer if we use the full object, at the expense of looking more unwieldy:
```
# Clone block and exclude attributes that do not support copy
cloneBlock( block, {}, null, {
	__experimentalExcludeAttributes: {
		__experimentalSupports: { copy: false }, # Here it's clear 'false' doesn't refer to exclusion
	}
);
```

I've also considered just passing a boolean, something like:
```
cloneBlock( block, {}, null, {
	excludeNonCopyableAttributes: false
);
```

dmsnell commented:
Is there a problem with leaving the abstraction one-level deep and exposing `{ __experimentalSupports: { copy: false } }` on `cloneBlock()`? I believe this was one of my earlier recommendations as well since it prevents introducing two new concepts when only one is being created.

```ts
cloneBlock( block, {}, null, { __experimentalSupports: { copy: false } } );
```

In terms of cloning that seems like there's a reasonable semantic connection between attributes we carry over and those we leave behind, plus it eliminates the confusion that `exclude` or `include` terminology introduces.

stacimc commented:
> Is there a problem with leaving the abstraction one-level deep and exposing { __experimentalSupports: { copy: false } } on cloneBlock()?

My worry is it conflates the concepts of what the clone operation supports (what it looks like) versus what the attributes it includes and excludes support (what it's actually doing). To clarify:

```
cloneBlock( block, {}, null, { __experimentalSupports: { copy: false } } );
```

You could interpret this to say the clone operation itself should not support `copy`. I don't think it's obvious that means we'll remove attributes that _don't_ support copy; the double negative is still an underlying issue. You could easily read this to say _copyable_ attributes shouldn't be 'supported', ie should be excluded.

Or to put it another way, let's say we add additional support keys for attributes, and I would like to clone a block and exclude attributes that _do_ support `export` (ie are configured with `"__experimentalSupports": { "export": true}`. I would probably guess the right way to do that would be:

```
// I want to clone a block, but I don't want to support exportable attributes
cloneBlock( block, {}, null, { __experimentalSupports: { export: false } } );
```

But that does the exact opposite of what I want, it excludes attributes that have `export: false`.

Maybe that latter case is contrived, but I think it's semantically clearer to be fully explicit, as heavy as it is:
```
// Clone block and exclude attributes that do not support copy
cloneBlock( block, {}, null, {
	__experimentalExcludeAttributes: {
		__experimentalSupports: { copy: false },
	}
);
```

stacimc committed:
Clarify arguments for cloneBlock

stacimc commented:
I pushed this update:

```
cloneBlock( block, {}, null, {
	__experimentalExcludeAttributes: {
		__experimentalSupports: { copy: false },
	}
);
```

For the reasons I mentioned [here](https://github.com/WordPress/gutenberg/pull/38643#issuecomment-1077852146). I think this is at least clear, although I acknowledge it's heavy-handed.

I think we could go in circles on this one, so I'm perfectly happy to change the language if folks generally prefer a simpler option. Given that this is only being added to `cloneBlock` and not to the serializer as well, there's less potential for misunderstanding. The language here is hard to get right and I really appreciate all the input 😄

dmsnell commented:
It sure is a mess 😄 but maybe we'll eventually think of something without the confusion. This feels like a direct _clone_ of the problem with `.gitignore`.

One last check: was there a problem with the array-of-supports-names? I still find that at least easier since it's presence-or-absence instead of a boolean.

Did we consider `__experimentalIgnoreAttributesSupporting: [ 'copy' ]`? Thinking of that on the same lines as `.gitignore` (if we're going to entertain heavy-handed names).

stacimc commented:
> It sure is a mess 😄 but maybe we'll eventually think of something without the confusion. This feels like a direct _clone_ of the problem with `.gitignore`.

Ha, definitely! Thank you for sticking with me 😅 

> One last check: was there a problem with the array-of-supports-names? I still find that at least easier since it's presence-or-absence instead of a boolean.
>
> Did we consider `__experimentalIgnoreAttributesSupporting: [ 'copy' ]`? Thinking of that on the same lines as `.gitignore` (if we're going to entertain heavy-handed names).

I think it would have to be even slightly longer! `__experimentalIgnoreAttributesNotSupporting: ['copy']`, adding in the 'Not' 😅 This wouldn't support the hypothetical future scenario where you want to ignore attributes that _do_ support some particular key, but it's definitely clear -- actually, I think even clearer than `__experimentalExcludeAttributes: { __experimentalSupports: { copy: false } }`.

_Or_ we could take your earlier suggestion to invert the logic and adjust it to `__experimentalRetainAttributesSupporting: ['copy']`. I still worry a little that "include attributes that support copy" might not convey "_and exclude attributes that don't_", but that feels like much less of a concern than the double negative 🤷‍♀️ Using 'retain' instead of 'include' I think helps as well.

dmsnell commented:
> `__experimentalIgnoreAttributesNotSupporting: ['copy']`

as long as that is I feel like it's the clearest yet. it even reveals how deep the double-negations go, as I think I'm still getting lost knowing what the intentions are when I read these.

`retain` seems at least decent if we want something shorter and without `Not` in the name.

`__experimentalOnlyCloneAttributesSupporting: [ 'copy' ]` is another wholly-positive construction

stacimc commented:
> as long as that is I feel like it's the clearest yet. it even reveals how deep the double-negations go, as I think I'm still getting lost knowing what the intentions are when I read these.

Agreed 😄 

Random inspiration: how about something like `__experimentalRequiredSupportKeys: [ 'copy' ]` or just `__experimentalRequiredSupports: [ 'copy' ]`?

dmsnell commented:
> `__experimentalRequiredSupports: [ 'copy' ]`

I like this, but with one possible modification that we link it back to attributes since otherwise it's only evident to you and me that's the case. blocks, which are cloned, also have `supports`

`__experimentalRequiredAttributeSupports: [ 'copy' ]`
`__experimentalRequireAttributeSupports: [ 'copy' ]`

stacimc committed:
Rename argument

stonetim commented:
@stacimc just ran into this issue and came across this PR looking for a solution. From what I can tell, it looks like this was just about ready to go and then stalled out a couple years ago. Has there been movement outside this PR or is there something else we can do to get the conflicts resolved and get this merged? Happy to assist.
